Failure to build without also explicitly adding serde as dependency

Every invocation of bit_struct! and enums! fails because of this:
```rust
error[E0463]: can't find crate for `serde`
  --> src/leaf6.rs:6:1
   |
6  | / bit_struct! {
7  | |
8  | |     pub struct ThermalFeatures(u32) {
9  | |         DTS: bool,
...  |
27 | |     }
28 | | }
   | |_^ can't find crate
   |
   = note: this error originates in the derive macro `$crate::serde::Deserialize` which comes from the expansion of the macro `bit_struct` (in Nightly builds, run with -Z macro-backtrace for more info)
```

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -11,12 +11,17 @@ keywords = ["bit", "struct", "macros"]
 categories = ["no-std"]
 rust-version = "1.62.1"
 
-# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
-
 [dependencies]
 num-traits = { version = "0.2", default-features = false }
-serde = { version = "1.0", default-features = false, features = ["derive"] }
+serde = { version = "1.0", default-features = false, features = ["derive"], optional = true }
 
 [dev-dependencies]
 trybuild = "1.0"
 matches = "0.1.9"
+quickcheck = "1.0"
+quickcheck_macros = "1.0"
+serde_json = "1.0"
+postcard = {version = "1.0.4", features = ["alloc"] }
+
+[features]
+default = ["serde"]

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -13,6 +13,7 @@ use core::{
 use num_traits::{Bounded, Num, One, Zero};
 /// Import serde here so we can reference it inside macros
 #[doc(hidden)]
+#[cfg(feature = "serde")]
 pub use serde;
 
 mod types;
@@ -370,6 +371,127 @@ macro_rules! bit_struct_impl {
     };
 }
 
+/// `serde` feature is not provided, so don't implement it
+#[doc(hidden)]
+#[macro_export]
+#[cfg(not(feature = "serde"))]
+macro_rules! bit_struct_serde_impl {
+    (
+        $(#[$meta:meta])*
+        $struct_vis:vis struct
+        $name:ident($kind:ty) { $($(#[$field_meta:meta])* $field:ident : $actual:ty),* $(,)? }
+    ) => {};
+}
+/// `serde` feature is provided, so implement it
+#[doc(hidden)]
+#[macro_export]
+#[cfg(feature = "serde")]
+macro_rules! bit_struct_serde_impl {
+    (
+        $(#[$meta:meta])*
+        $struct_vis: vis struct $name: ident ($kind: ty) {
+        $(
+            $(#[$field_meta:meta])*
+            $field: ident: $actual: ty
+        ),* $(,)?
+        }
+    ) => {
+        #[allow(clippy::used_underscore_binding)]
+        impl $crate::serde::Serialize for $name {
+            fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error> where S: $crate::serde::Serializer {
+                use $crate::serde::ser::SerializeStruct;
+
+                let mut v = *self;
+
+                let mut serializer = serializer.serialize_struct(
+                    stringify!($name),
+                    $crate::count_idents!( 0, [$( $field ),*] ),
+                )?;
+                $(
+                    serializer.serialize_field(
+                        stringify!($field),
+                        &v.$field().get()
+                    )?;
+                )*
+                serializer.end()
+            }
+        }
+
+        #[allow(clippy::used_underscore_binding)]
+        impl<'de> $crate::serde::Deserialize<'de> for $name {
+            fn deserialize<D>(deserializer: D) -> Result<Self, D::Error> where D: $crate::serde::Deserializer<'de> {
+
+                use $crate::serde::de::{self, Deserialize, Deserializer, MapAccess, SeqAccess, Visitor};
+                use ::core::fmt;
+
+                const FIELDS: &'static [&'static str] = &[ $( stringify!( $field ) ),* ];
+
+                #[allow(non_camel_case_types)]
+                enum Fields { $( $field ),* }
+                impl<'de> Deserialize<'de> for Fields {
+                    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
+                        struct FieldVisitor;
+                        impl<'de> Visitor<'de> for FieldVisitor {
+                            type Value = Fields;
+
+                            fn expecting(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+                                f.write_str(stringify!( $( $field ),* ))
+                            }
+
+                            fn visit_str<E: de::Error>(self, value: &str) -> Result<Fields, E> {
+                                match value {
+                                    $( stringify!( $field ) => Ok(Fields::$field), )*
+                                    _ => Err(de::Error::unknown_field(value, FIELDS)),
+                                }
+                            }
+                        }
+
+                        deserializer.deserialize_identifier(FieldVisitor)
+                    }
+                }
+
+                struct BitStructVisitor;
+                impl<'de> Visitor<'de> for BitStructVisitor {
+                    type Value = $name;
+
+                    fn expecting(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+                        f.write_str(concat!("struct ", stringify!($name)))
+                    }
+
+                    fn visit_map<V: MapAccess<'de>>(self, mut map: V) -> Result<$name, V::Error> {
+                        $( let mut $field: Option<$actual> = None; )*
+                        while let Some(key) = map.next_key::<Fields>()? {
+                            match key {
+                                $( Fields::$field => {
+                                    if $field.is_some() {
+                                        return Err(de::Error::duplicate_field(stringify!($field)));
+                                    }
+                                    $field = Some(map.next_value()?);
+                                },)*
+                            }
+                        }
+                        $(
+                            let $field = $field.ok_or_else(|| de::Error::missing_field(stringify!($field)))?;
+                        )*
+                        Ok($name::new( $( $field ),* ))
+                    }
+
+                    fn visit_seq<V: SeqAccess<'de>>(self, mut seq: V) -> Result<$name, V::Error> {
+                        let mut count = 0;
+                        $(
+                            let $field = seq.next_element()?
+                                .ok_or_else(|| de::Error::invalid_length(count, &self))?;
+                            count += 1;
+                        )*
+                        Ok($name::new( $( $field ),* ))
+                    }
+                }
+                deserializer.deserialize_struct(stringify!($name), FIELDS, BitStructVisitor)
+            }
+        }
+    }
+}
+
 /// A bit struct which has a zero value we can get
 pub trait BitStructZero: Zero {
     /// Get a zero value for this bit struct
@@ -487,38 +609,13 @@ macro_rules! bit_struct {
         #[derive(Copy, Clone, PartialOrd, PartialEq, Eq, Ord, Hash)]
         pub struct $name($crate::UnsafeStorage<$kind>);
 
-        #[allow(clippy::used_underscore_binding)]
-        impl $crate::serde::Serialize for $name {
-            fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error> where S: $crate::serde::Serializer {
-
-                let mut v = *self;
-                #[derive($crate::serde::Serialize)]
-                struct Raw {
-                    $($field: $actual),*
-                }
-
-                let raw = Raw {
-                    $($field: v.$field().get(),)*
-                };
-
-                raw.serialize(serializer)
-            }
-        }
-
-        #[allow(clippy::used_underscore_binding)]
-        impl $crate::serde::Deserialize<'static> for $name {
-            fn deserialize<D>(deserializer: D) -> Result<Self, D::Error> where D: $crate::serde::Deserializer<'static> {
-
-                #[derive($crate::serde::Deserialize)]
-                struct Raw {
-                    $($field: $actual),*
-                }
-
-                let raw = Raw::deserialize(deserializer)?;
-
-                Ok($name::new(
-                    $(raw.$field),*
-                ))
+        $crate::bit_struct_serde_impl! {
+            $(#[$meta])*
+            $struct_vis struct $name ($kind) {
+            $(
+                $(#[$field_meta])*
+                $field: $actual
+            ),*
             }
         }
 
@@ -593,7 +690,7 @@ macro_rules! count_idents {
     };
     ($on: expr, []) => {
         $on
-    }
+    };
 }
 
 /// Returns the index of the leading 1 in `num`
@@ -620,6 +717,113 @@ pub const fn bits(num: usize) -> usize {
     }
 
     helper(0, num)
+}
+
+/// `serde` feature is not provided, so don't implement it
+#[doc(hidden)]
+#[cfg(not(feature = "serde"))]
+#[macro_export]
+macro_rules! enum_serde_impl {
+    ($enum_vis:vis $name:ident { $fst_field:ident $(, $field:ident)* }) => {};
+}
+
+/// `serde` feature is provided, so implement it
+#[doc(hidden)]
+#[cfg(feature = "serde")]
+#[macro_export]
+macro_rules! enum_serde_impl {
+    ($name:ident { $($field:ident),* }) => {
+        impl $crate::serde::Serialize for $name {
+            fn serialize<S: $crate::serde::Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+                match self {
+                    $(
+                        Self::$field => {
+                            serializer.serialize_unit_variant(
+                                stringify!($name),
+                                *self as u32,
+                                stringify!($field),
+                            )
+                        },
+                    )*
+                }
+            }
+        }
+        impl<'de> $crate::serde::Deserialize<'de> for $name {
+            fn deserialize<D>(deserializer: D) -> Result<Self, D::Error> where D: $crate::serde::Deserializer<'de> {
+                use ::core::{fmt, result::Result::{self, Ok, Err}, convert::TryFrom};
+                use $crate::serde::de::{Deserialize, Deserializer, EnumAccess, VariantAccess, Visitor};
+
+                #[repr(u64)]
+                enum Variants { $( $field ),* }
+                impl TryFrom<u64> for Variants {
+                    type Error = ();
+
+                    fn try_from(v: u64) -> Result<Self, Self::Error> {
+                        if v < $crate::count_idents!(0, [$( $field ),*]) {
+                            // SAFETY:
+                            // This is safe because we're converting a `u64` to a `repr(u64)`
+                            // enum, and we've checked that the value is one of the variants.
+                            unsafe { core::mem::transmute(v) }
+                        } else {
+                            Err(())
+                        }
+                    }
+                }
+                impl<'de> Deserialize<'de> for Variants {
+                    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
+                        struct VariantsVisitor;
+                        impl<'de> Visitor<'de> for VariantsVisitor {
+                            type Value = Variants;
+                            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                                formatter.write_str("variant identifier")
+                            }
+
+                            fn visit_u64<E: $crate::serde::de::Error>(self, value: u64) -> Result<Self::Value, E> {
+                                Variants::try_from(value)
+                                    .map_err(|()| $crate::serde::de::Error::invalid_value(
+                                        $crate::serde::de::Unexpected::Unsigned(value),
+                                        &"variant index"
+                                    ))
+                            }
+
+                            fn visit_str<E: $crate::serde::de::Error>(self, value: &str) -> Result<Self::Value, E> {
+                                match value {
+                                    $( stringify!($field) => Ok(Variants::$field), )*
+                                    _ => Err($crate::serde::de::Error::unknown_variant(value, VARIANTS)),
+                                }
+                            }
+                        }
+                        deserializer.deserialize_identifier(VariantsVisitor)
+                    }
+                }
+
+                struct EnumVisitor;
+                impl<'de> Visitor<'de> for EnumVisitor {
+                    type Value = $name;
+                    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                        formatter.write_str(concat!("enum ", stringify!($name)))
+                    }
+
+                    fn visit_enum<A: EnumAccess<'de>>(self, data: A) -> Result<Self::Value, A::Error> {
+                        match data.variant()? {
+                            $(
+                            (Variants::$field, variant) => {
+                                let () = variant.unit_variant()?;
+                                Ok($name::$field)
+                            }
+                            ),*
+                        }
+                    }
+                }
+                const VARIANTS: &'static [&'static str] = &[ $( stringify!( $field ) ),* ];
+                deserializer.deserialize_enum(
+                    stringify!($name),
+                    VARIANTS,
+                    EnumVisitor,
+                )
+            }
+        }
+    };
 }
 
 /// Helper macro
@@ -699,7 +903,7 @@ macro_rules! enum_impl {
     ) => {
         #[repr(u8)]
         $(#[$meta])*
-        #[derive(Copy, Clone, Debug, PartialOrd, PartialEq, Eq, $crate::serde::Serialize, $crate::serde::Deserialize)]
+        #[derive(Copy, Clone, Debug, PartialOrd, PartialEq, Eq)]
         $enum_vis enum $name {
             $(#[$fst_field_meta])*
             $fst_field,
@@ -708,6 +912,8 @@ macro_rules! enum_impl {
                 $field
             ),*
         }
+
+        $crate::enum_serde_impl! { $name { $fst_field $(, $field)* } }
 
         unsafe impl $crate::BitCount for $name {
             const COUNT: usize = $crate::bits($crate::count_idents!(0, [$($field),*]));
@@ -752,7 +958,7 @@ macro_rules! enum_impl {
     ) => {
         #[repr(u8)]
         $(#[$meta])*
-        #[derive(Copy, Clone, Debug, PartialOrd, PartialEq, Eq, $crate::serde::Serialize, $crate::serde::Deserialize)]
+        #[derive(Copy, Clone, Debug, PartialOrd, PartialEq, Eq)]
         $enum_vis enum $name {
             $(#[$fst_field_meta])*
             $fst_field,
@@ -761,6 +967,8 @@ macro_rules! enum_impl {
                 $field
             ),*
         }
+
+        $crate::enum_serde_impl! { $name { $fst_field $(, $field)* } }
 
         impl Default for $name {
             fn default() -> Self {

--- a/src/types.rs
+++ b/src/types.rs
@@ -2,6 +2,7 @@
 //! types
 
 use super::*;
+#[cfg(feature = "serde")]
 use serde::{Deserializer, Serializer};
 
 /// Assert that the given type is valid for any representation thereof
@@ -84,6 +85,7 @@ macro_rules! new_signed_types {
 
         always_valid!($name);
 
+        #[cfg(feature = "serde")]
         impl serde::Serialize for $name {
             fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
             where
@@ -93,6 +95,7 @@ macro_rules! new_signed_types {
             }
         }
 
+        #[cfg(feature = "serde")]
         impl <'de> serde::Deserialize<'de> for $name {
             fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
             where
@@ -406,6 +409,7 @@ macro_rules! new_unsigned_types {
             }
         }
 
+        #[cfg(feature = "serde")]
         impl serde::Serialize for $name {
             fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
             where
@@ -415,6 +419,7 @@ macro_rules! new_unsigned_types {
             }
         }
 
+        #[cfg(feature = "serde")]
         impl <'de> serde::Deserialize<'de> for $name {
             fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
             where

--- a/tests/serialization.rs
+++ b/tests/serialization.rs
@@ -1,0 +1,77 @@
+#![cfg(feature = "serde")]
+
+use quickcheck::{Arbitrary, Gen};
+
+bit_struct::enums! {
+    pub Color { Orange, Red, Blue, Yellow, Green }
+}
+
+bit_struct::bit_struct! {
+    struct BitStruct(u32) {
+        a_color: Color,
+        b: bit_struct::u3,
+    }
+}
+
+impl Arbitrary for Color {
+    fn arbitrary(g: &mut Gen) -> Self {
+        *g.choose(&[
+            Self::Orange,
+            Self::Red,
+            Self::Blue,
+            Self::Yellow,
+            Self::Green,
+        ])
+        .unwrap()
+    }
+}
+
+impl Arbitrary for BitStruct {
+    fn arbitrary(g: &mut Gen) -> Self {
+        let b = *g
+            .choose(&[
+                bit_struct::u3!(0),
+                bit_struct::u3!(1),
+                bit_struct::u3!(2),
+                bit_struct::u3!(3),
+                bit_struct::u3!(4),
+                bit_struct::u3!(5),
+                bit_struct::u3!(6),
+                bit_struct::u3!(7),
+            ])
+            .unwrap();
+        Self::new(Color::arbitrary(g), b)
+    }
+}
+
+#[quickcheck_macros::quickcheck]
+fn test_round_trip_serialize_json_enum(color: Color) -> bool {
+    use serde_json::{from_value, to_value};
+    let round_trip: Color =
+        from_value(to_value(color).expect("Failed to serialize")).expect("Failed to deserialize");
+    round_trip == color
+}
+
+#[quickcheck_macros::quickcheck]
+fn test_round_trip_serialize_postcard_enum(color: Color) -> bool {
+    use postcard::{from_bytes, to_allocvec};
+    let round_trip: Color = from_bytes(&to_allocvec(&color).expect("Failed to serialize"))
+        .expect("Failed to deserialize");
+    round_trip == color
+}
+
+#[quickcheck_macros::quickcheck]
+fn test_round_trip_serialize_json_struct(bits: BitStruct) -> bool {
+    use serde_json::{from_value, to_value};
+    let round_trip: BitStruct =
+        from_value(to_value(bits).expect("Failed to serialize")).expect("Failed to deserialize");
+    round_trip == bits
+}
+
+#[quickcheck_macros::quickcheck]
+fn test_round_trip_serialize_postcard_struct(bits: BitStruct) -> bool {
+    use postcard::{from_bytes, to_allocvec};
+    let round_trip: BitStruct = from_bytes(&to_allocvec(&bits).expect("Failed to serialize"))
+        .expect("Failed to deserialize");
+    round_trip == bits
+}

